nomenclature/processor/meta.py

```diff
@@ -0,0 +1,89 @@
+from pathlib import Path
+import pyam
+from nomenclature.processor import Processor
+from nomenclature.codelist import MetaCodeList
+
+
+class MetaValidator(Processor):
+    """Meta indicator validation and processing class"""
+
+    meta_code_list: MetaCodeList = None
+
+    def __init__(self, path_to_meta_code_list_files: Path):
+        super().__init__()
+        self.meta_code_list = MetaCodeList.from_directory(
+            name="meta_code_list", path=path_to_meta_code_list_files
+        )
+
+    def _values_allowed(self, values, allowed_values, meta_indicator) -> bool:
+        """Checks if the values within a meta indicator column are
+        listed in model mapping
+
+        Parameters
+        ----------
+        values :
+            List of values in the meta_indicator column of the df: IamDataFrame.
+        allowed_values :
+            List of allowed values for the meta_indicator column
+        meta_indicator :
+            The name of the meta_indicator/column whose values are being checked.
+
+        Returns
+        -------
+        True : boolean
+            If all column elements are listed in model mapping
+
+        Raises
+        ------
+        ValueError
+            *If any of the values in the meta indicator column are not
+            listed in model mapping
+
+
+        """
+        not_allowed = [value for value in values if value not in allowed_values]
+        if not_allowed:
+            raise ValueError(
+                f"{not_allowed} meta indicator value(s) in the {meta_indicator} "
+                "column are not allowed. Allowed values are "
+                f"{allowed_values}"
+            )
+        return True
+
+    def apply(self, df: pyam.IamDataFrame) -> pyam.IamDataFrame:
+        """Apply meta indicator validation processing
+
+        Parameters
+        ----------
+        df (pyam.IamDataFrame)
+            Input data whose meta indicators will be validated
+
+        Returns
+        -------
+        df (pyam.IamDataFrame)
+            If all meta indicators and their values are listed in the
+            model mapping, the same df is returned.
+
+        Raises
+        ------
+        ValueError
+            *If a meta indicator in the 'df' is not listed in the .yaml
+            definition file
+        """
+        unrecognized_meta_indicators = []
+        for meta_indicator in df.meta.columns:
+            if meta_indicator not in self.meta_code_list.mapping:
+                unrecognized_meta_indicators.append(meta_indicator)
+            else:
+                values = df.meta[meta_indicator].values
+                allowed_values = self.meta_code_list.mapping[
+                    meta_indicator
+                ].allowed_values
+                self._values_allowed(values, allowed_values, meta_indicator)
+        if unrecognized_meta_indicators:

```

Comment:
I would swap the order of things, because right now you:
1. Check if there's any unrecognized meta indicators
2. Check if there's any values that are not allowed
  2.1. If there are values that are not allowed, you raise an error
3. If there are unrecognized meta indicators you raise an error

My suggestion would be to split the two steps more cleanly like so:

```python
# check for illegal columns
if unrecognized_meta_indicators := [
    meta_indicator
    for meta_indicator in df.meta.colums
    if meta_indicator not in self.meta_code_list.mapping
]:
    raise ValueError(
        f"{unrecognized_meta_indicators} is/are not recognized in the "
        f"meta definitions file. Allowed meta indicators are: "
        f"{list(self.meta_code_list.mapping)}"
    )
# check for illegal values
for meta_indicator in df.meta.columns:
    self._values_allowed(
        df.meta[meta_indicator].unique,
        self.meta_code_list.mapping[meta_indicator].allowed_values,
        meta_indicator,
    )
```